astropy/utils/collections.py

```diff
@@ -54,3 +56,46 @@ def extend(self, x):
         for item in x:
             self._assert(item)
             super().append(item)
+
+
+class _DisplayFormatAttributeMixin:
```

Comment:
I'm not sure where this should live (this location feels wrong), but don't have a better suggestion beyond creating a new module in `astropy/utils`. I'm also wondering why the "_" preface -- if this is in utils, it is sort of implied that it's an internal thing, so maybe it'd be fine to just call it `DisplayFormatAttributeMixin`?

Reply:
I also dislike having both the ``fmt`` and ``format``. The former refers to the ``format_spec``, e.g. "0.3g" in python's format mini-language. The latter is for displaying in latex versus html, etc.
I guess I can deprecate the ``fmt`` argument in favor of the more verbose``format_spec``. It's still a bit confusing.
Perhaps the ``format`` kwarg should likewise be changed, IDK to what.

Reply:
> if this is in utils, it is sort of implied that it's an internal thing, so maybe it'd be fine to just call it `DisplayFormatAttributeMixin`

SGTM. It was mostly because this is very specific to the two use cases. I'm not sure if requiring a ``.name`` attribute is sensical for a general mixin. If we were to make this public, I'd probably change this to ``_format_fallback``, itself falling back to ``.name`` or something.

Reply:
So I'm still on the fence about making this public, given the need for a ``.name`` attribute.

Reply:
The API docstring will be incomplete if this is hidden, as the docstring here won't be rendered.

If this is going to be in utils and used in two different subpackages, I think it should be public.

Reply:
SGTM. Then I think we should address the requirement for a ``.name`` attribute.

I see a few options:
1. leave it as a requirement

```python
    return self._format.get(format, self.name)
```

This punts the question, but does commit us to leaving ``self.name`` as the final fallback if the format can't be found.

3. drop the requirement. If it's an attribute on the class use it, else error.
```python
    if hasattr(self, "name"):
        return self._format.get(format, self.name)
    else:
        return self._format[format]
```

3. drop the requirement. If it's an attribute on the class use it, else return None.

```python
    return self._format.get(format, getattr(self, "name", None))
```

4. drop the requirement. Add an fallback value variable. If the fallback variable is not defined, we can error, return None, or try to get the name...

```python
    fallback = getattr(self, "_fallback_format")   # can error, get None, or try to get the name...
    return self._format.get(format, )
```

Reply:
@pllim, I think 

```python
fallback = getattr(self, "_fallback_format", self.name)   # can error, get None, or try to get the name...
return self._format.get(format, fallback)
```

is the best option, but we don't need to implement that now, as currently all classes using this mixin class define a ``name`` attribute.

Reply:
Since this is also used in `units`, do @mhvk or @adrn have opinions on which one makes most sense given its purpose?

Reply:
I see this as a question of orthogonality. How much do we want this mixin to overlap with common structures, like a ``name`` attribute? In particular, many classes, like Cosmology allow ``name`` to be `None`, which is less compatible with how this mixin operates. This isn't to say that falling back to ``.name`` is bad, it just adds complexity when dealing with non-string edge cases. OTOH, adding completely orthogonal requirements like ``_fallback_format`` can cause code duplication. 🤷 

I think this mixin *might* be useful in some other places — `Constant` and `Cosmology` off the top of my head.

Reply:
@adrn @mhvk, I'd like to get this PR in soon, so #12356 (by a first time contributor) can proceed. Which of the four above options do you all think is best. How general do we want to go with this mixin?

Reply:
I don't have a strong opinion. I like the compromise approach of
```python
    return self._format.get(format, getattr(self, "name", None))
```
in that it's close enough to the original behavior in `NamedUnit`, but makes intuitive sense to me to support `None` when `.name` is not available. But I think @mhvk has wrapped his mind around how this is needed in astropy.units more recently...